File: lib/free_llm_client.py
"""
Free LLM API Client (Pekpik)

Usage:
    from lib.free_llm_client import chat, generate_image, list_models

    # Chat with Claude Opus
    response = chat("Write a script for a 60s promo")

    # Generate image with DALL-E 3
    image_url = generate_image("A vinyl album cover for electronic music")

    # List available models
    models = list_models()
"""

import os
from typing import Optional, Dict, Any, List
from openai import OpenAI

PEKPIK_BASE = os.getenv("PEKPIK_BASE_URL", "https://aiapiv2.pekpik.com/v1")
PEKPIK_KEY = os.getenv("PEKPIK_API_KEY", "sk-placeholder")

MODELS = {
    "CLAUDE_OPUS": "claude-opus-4-7",
    "GPT_5_5": "openai/gpt-5.5",
    "GEMINI": "gemini-2.5-flash",
    "DEEPSEEK": "deepseek-v4-flash",
}

# Initialize client
client = OpenAI(
    base_url=PEKPIK_BASE,
    api_key=PEKPIK_KEY,
)


def chat(
    prompt: str,
    model: str = MODELS["CLAUDE_OPUS"],
    temperature: float = 0.7,
    max_tokens: int = 2000,
) -> str:
    """Chat completion with free LLM."""
    try:
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            temperature=temperature,
            max_tokens=max_tokens,
        )
        return response.choices[0].message.content
    except Exception as e:
        print(f"[Pekpik] Chat failed ({model}): {e}")
        raise


def generate_image(
    prompt: str,
    size: str = "1024x1024",
    quality: str = "standard",
) -> str:
    """Generate image with DALL-E 3."""
    try:
        response = client.images.generate(
            model="dall-e-3",
            prompt=prompt,
            size=size,
            quality=quality,
            n=1,
        )
        return response.data[0].url
    except Exception as e:
        print(f"[Pekpik] Image generation failed: {e}")
        raise


def text_to_speech(
    text: str,
    voice: str = "alloy",
) -> bytes:
    """Text-to-speech with TTS-1-HD."""
    try:
        response = client.audio.speech.create(
            model="tts-1-hd",
            voice=voice,
            input=text,
        )
        return response.content
    except Exception as e:
        print(f"[Pekpik] TTS failed: {e}")
        raise


def embed(
    texts: List[str] | str,
) -> List[Dict[str, Any]]:
    """Generate embeddings with text-embedding-3-small."""
    input_texts = texts if isinstance(texts, list) else [texts]
    try:
        response = client.embeddings.create(
            model="text-embedding-3-small",
            input=input_texts,
        )
        return response.data
    except Exception as e:
        print(f"[Pekpik] Embeddings failed: {e}")
        raise


def list_models() -> Dict[str, str]:
    """List available models."""
    return MODELS.copy()

File: lib/free-llm-client.js
/**
 * Free LLM API Client (Pekpik)
 *
 * Usage:
 *   const { chat, generateImage } = require('./free-llm-client');
 *
 *   // Chat with Claude Opus
 *   const response = await chat("Write a script for a 60s promo", { model: 'claude-opus-4-7' });
 *
 *   // Generate image with DALL-E 3
 *   const image = await generateImage("A vinyl album cover for electronic music");
 */

const OpenAI = require('openai').default;

const PEKPIK_BASE = process.env.PEKPIK_BASE_URL || 'https://aiapiv2.pekpik.com/v1';
const PEKPIK_KEY = process.env.PEKPIK_API_KEY;

const MODELS = {
  CLAUDE_OPUS: 'claude-opus-4-7',
  GPT_5_5: 'openai/gpt-5.5',
  GEMINI: 'gemini-2.5-flash',
  DEEPSEEK: 'deepseek-v4-flash',
};

// Initialize client
const client = new OpenAI({
  baseURL: PEKPIK_BASE,
  apiKey: PEKPIK_KEY || 'sk-placeholder',
});

/**
 * Chat completion
 */
async function chat(prompt, options = {}) {
  const {
    model = MODELS.CLAUDE_OPUS,
    temperature = 0.7,
    maxTokens = 2000,
  } = options;

  try {
    const response = await client.chat.completions.create({
      model,
      messages: [{ role: 'user', content: prompt }],
      temperature,
      max_tokens: maxTokens,
    });

    return response.choices[0].message.content;
  } catch (error) {
    console.error(`[Pekpik] Chat failed (${model}):`, error.message);
    throw error;
  }
}

/**
 * Generate image with DALL-E 3
 */
async function generateImage(prompt, options = {}) {
  const { size = '1024x1024', quality = 'standard' } = options;

  try {
    const response = await client.images.generate({
      model: 'dall-e-3',
      prompt,
      size,
      quality,
      n: 1,
    });

    return response.data[0].url;
  } catch (error) {
    console.error('[Pekpik] Image generation failed:', error.message);
    throw error;
  }
}

/**
 * Text-to-speech with TTS-1-HD
 */
async function textToSpeech(text, options = {}) {
  const { voice = 'alloy' } = options;

  try {
    const response = await client.audio.speech.create({
      model: 'tts-1-hd',
      voice,
      input: text,
    });

    return response;
  } catch (error) {
    console.error('[Pekpik] TTS failed:', error.message);
    throw error;
  }
}

/**
 * Embeddings with text-embedding-3-small
 */
async function embed(texts, options = {}) {
  const input = Array.isArray(texts) ? texts : [texts];

  try {
    const response = await client.embeddings.create({
      model: 'text-embedding-3-small',
      input,
    });

    return response.data;
  } catch (error) {
    console.error('[Pekpik] Embeddings failed:', error.message);
    throw error;
  }
}

/**
 * List available models
 */
function listModels() {
  return Object.entries(MODELS).reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {});
}

module.exports = {
  chat,
  generateImage,
  textToSpeech,
  embed,
  listModels,
  MODELS,
  PEKPIK_BASE,
};
